nomad/secure_variables_endpoint.go

```diff
@@ -1,133 +1,470 @@
+	case nil:
+		// If ACLs are enabled, ensure the caller has the submit-job namespace
+		// capability.
+		if aclObj != nil {
+			hasSubmitJob := aclObj.AllowNsOp(args.RequestNamespace(), acl.NamespaceCapabilitySubmitJob)
+			if !hasSubmitJob {
+				return structs.ErrPermissionDenied
+			}
+		}
+	default:
+		// This endpoint is generally called by Nomad nodes, so we want to
+		// perform this check, unless the token resolution gave us a terminal
+		// error.
+		if err != structs.ErrTokenNotFound {
+			return err
+		}
+
+		// Attempt to lookup AuthToken as a Node.SecretID and return any error
+		// wrapped along with the original.
+		node, stateErr := sv.srv.fsm.State().NodeBySecretID(nil, args.AuthToken)
+		if stateErr != nil {
+			var mErr multierror.Error
+			mErr.Errors = append(mErr.Errors, err, stateErr)
+			return mErr.ErrorOrNil()
+		}
+
+		// At this point, we do not have a valid ACL token, nor are we being
+		// called, or able to confirm via the state store, by a node.
+		if node == nil {
+			return structs.ErrTokenNotFound
+		}
 	}
 
-	sv.logger.Trace("TODO") // silences structcheck lint
+	// Use a multierror, so we can capture all validation errors and pass this
+	// back so fixing in a single swoop.
+	var mErr multierror.Error
 
-	// TODO: placeholder for serialization and encryption
-	var buf bytes.Buffer
-	enc := gob.NewEncoder(&buf)
-	err := enc.Encode(args.Data.UnencryptedData)
-	if err != nil {
+	// Iterate the secure variables and validate them. Any error results in the
+	// call failing.
+	for _, i := range args.Data {
+		if err := i.Validate(); err != nil {
+			mErr.Errors = append(mErr.Errors, err)
+		}
+		if err := sv.encrypt(i); err != nil {
+			mErr.Errors = append(mErr.Errors, err)
+		}
+	}
+	if err := mErr.ErrorOrNil(); err != nil {
 		return err
 	}
 
-	args.Data.EncryptedData = &structs.SecureVariableData{}
-	args.Data.EncryptedData.KeyID = "TODO"
-	args.Data.EncryptedData.Data = sv.encrypter.Encrypt(buf.Bytes(), args.Data.EncryptedData.KeyID)
+	// Update via Raft.
+	out, index, err := sv.srv.raftApply(structs.SecureVariableUpsertRequestType, args)
+	if err != nil {
+		return err
+	}
 
-	// TODO: implementation
-	SV_Upsert(args, reply)
+	// Check if the FSM response, which is an interface, contains an error.
+	if err, ok := out.(error); ok && err != nil {
+		return err
+	}
 
+	// Update the index. There is no need to floor this as we are writing to
+	// state and therefore will get a non-zero index response.
+	reply.Index = index
 	return nil
 }
 
-func (sv *SecureVariables) List(args *structs.SecureVariablesListRequest, reply *structs.SecureVariablesListResponse) error {
-	if done, err := sv.srv.forward("SecureVariables.List", args, args, reply); done {
+// Delete removes a single secure variable, as specified by its namespace and
+// path from Nomad. This is typically called by Nomad nodes, however, in extreme
+// situations can be used via the CLI and API by operators.
+func (sv *SecureVariables) Delete(
```

Comment:
I'm not so sure about this comment and associated ACL workflow... when is `SecureVariables.Delete` going to be called by Nomad nodes? Aren't all the `SecureVariables.*` RPCs driven by either humans or by the template implementation (which is indirectly through humans too)?

Same applies for `SecureVariables.Upsert` and `handleMixedAuthEndpoint`